Add VHeaderInternal

obulat opened the pull request:
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, please consider opening one before creating this pull request. -->
Fixes #1807 by @obulat
Fixes #1806 by obulat

## Description
<!-- Concisely describe what the pull request does. -->
<!-- Add screenshots, videos, or other media to show the problem and the solution when appropriate. -->
This PR creates a Content header with the links to internal pages. I copied a large part of the implementation from the VFooter PR (#1814).

For this PR, I had to also add the following changes to the `VModal` component:
- Make the VModal controllable from outside by adding the `externalVisible` prop which is used in addition to the `visible` ref inside the component to manage the modal visibility.
- expose the `top-bar` slot from `VModalContent` to the `VModal` consumers.
- add `mode` (light/dark) and `variant` (default/full) props to make the modal content stylable.

## Testing Instructions
<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete this section entirely. -->

1. Run `pnpm dev`.
2. Go to localhost:8443/preferences and check the `new_header` flag
3. Go to localhost:8443/about
4. You should see the new Content header. Check that it works as expected on all breakpoints and matches the [Figma mockups](https://www.figma.com/file/KxkZmBNk218V2Fke1GeYCu/New-header-component?node-id=609%3A70088).

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. -->
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

obulat commented on src/composables/use-pages.ts:
```diff
@@ -47,7 +47,7 @@ export default function usePages() {
 
   const route = useRoute()
-  const currentPageId = computed(() => route.value.name)
+  const currentPageId = computed(() => route.value?.name?.split('__')[0])
```
Originally, this returned the i18n name like `about__en`, which would not match the ids in the `pages` array.

sarayourfriend commented:
@obulat Is the name of this PR accurate still? There's no new component called VHeaderInternal in this PR :thinking:

obulat commented:
> @obulat Is the name of this PR accurate still? There's no new component called VHeaderInternal in this PR :thinking: 

It was the name of the component that was inaccurate, @sarayourfriend. I realized it after rereading the issue, and updated the name of the component.

obulat commented on src/components/VBrand/VBrand.vue:
```diff
@@ -0,0 +1,44 @@
+    srText: {
+      type: String,
+      default: 'Openverse',
+    },
```
This was copied from the VFooter PR, which was merged in after I started working on this issue.

sarayourfriend replied:
Which part was copied? And why does it appear in this PR if it's merged on main? Does this PR need to be rebased then?

obulat commented on src/components/VModal/VModal.vue:
```diff
@@ -104,6 +110,22 @@ export default defineComponent({
+      ),
+      default: 'light',
+    },
+    externalVisible: {
+      type: Boolean,
+      default: undefined,
+    },
```
This is an excellent suggestion, thank you!

obulat commented on src/components/VModal/VModal.vue:
```diff
@@ -132,6 +157,15 @@ export default defineComponent({
+        lock()
+      } else {
+        unlock()
+      }
```
> Were you trying to intentionally avoid emitting the event, by chance?

I was, and I was wrong. Thank you for pointing this out!

sarayourfriend commented on src/components/VHeader/VHeaderInternal.vue:
```diff
@@ -0,0 +1,137 @@
+            <VBrand :is-fetching="false" />
+            <VIconButton
+              ref="closeButton"
+              variant="plain"
```
This does not appear to be an actual prop on `VIconButton`. Is this serving a different function here?

```suggestion
```

sarayourfriend reviewed (dismissed):
~This renders with the links vertically misaligned compared to the brand for me:~
![image](https://user-images.githubusercontent.com/24264157/191395512-a954d364-3d35-40f7-ba47-e8507465f999.png)

Edit: I was on an old version of the branch, sorry. The latest version of the branch does not have misaligned links.

Can you add snapshot tests as well for each breakpoint, either of the about page itself using the query param toggle for the feature flag or in a storybook playwright test?

I am still reviewing the code a little more closely.

sarayourfriend commented:
A behaviour I just noticed is that the modal stays open if you resize the window back to a breakpoint bigger than when the modal would be able to be opened. I think this is better than disappearing the modal based on the breakpoint, but I wanted to make sure that behaviour is intentional. We may have to wait to ask Francisco once he is available though.

sarayourfriend commented on src/components/VHeader/VHeaderInternal.vue:
```diff
@@ -0,0 +1,137 @@
+import WordPress from '~/assets/wordpress.svg?inline'
+
+export default defineComponent({
+  name: 'VHeaderInternal',
```
I don't want to bike shed too much on the name of this component, but I don't really understand what "Internal" means here. Is it only meant for special, non-public pages or something?

obulat replied:
I think the 'Internal' here comes from the Internal pages that this header links to. The non-search and non-content-details header. Do you have any suggestions for the component name?

sarayourfriend commented on src/components/VHeader/VHeaderInternal.vue:
```diff
@@ -0,0 +1,137 @@
+      isModalVisible.value = visibility
+    }
+
+    // When clicking on an internal link in the modal, close the modal
+    watch(route, () => {
+      if (isModalVisible.value) {
+        setModalVisibility(false)
+      }
+    })
```
The modal closes when you click an external link as well. Is that intentional?

The modal also causes the page to scroll to the top now instead of staying at the same scroll position. This appears to be a regression in this PR as it does not do it on the staging deployment.

Both behaviours are illustrated in the video below.


https://user-images.githubusercontent.com/24264157/191396993-9fdac266-f8d6-48c9-8855-5bb4540c83ba.mp4

zackkrida replied:
> The modal also causes the page to scroll to the top now instead of staying at the same scroll position. This appears to be a regression in this PR as it does not do it on the staging deployment.

Observing this too.

obulat replied:
I think I've fixed this in https://github.com/WordPress/openverse-frontend/pull/1825/commits/e445d9a14426ee58b9c911aaefbe77e2e8056159, are you still seeing this behavior, @zackkrida ?

sarayourfriend reviewed (commented):
In addition to snapshot tests for the new header, e2e tests exercising the modal opening closing, scroll locking (once fixed) and external link handling would also be good to have.

obulat commented:
> A behaviour I just noticed is that the modal stays open if you resize the window back to a breakpoint bigger than when the modal would be able to be opened. I think this is better than disappearing the modal based on the breakpoint, but I wanted to make sure that behaviour is intentional. We may have to wait to ask Francisco once he is available though.

It wasn't intentional, fixing it.

zackkrida reviewed (changes requested):
This looks excellent, some things I noticed:

- The logo is not a clickable link, it's just a span
- The focus ring in the mobile menu includes a thin white ring around the pink one
- if you expand the page while the menu is open it stays open. not sure if this is worth fixing. Screenshot: https://share.cleanshot.com/NYBDBV

zackkrida commented on src/components/VHeader/VHeaderInternal.vue:
```diff
@@ -0,0 +1,150 @@
+      </VModal>
+    </div>
+  </header>
+</template>
+<script lang="ts">
```
```suggestion
</template>

<script lang="ts">
```

obulat commented:
I've extracted the logo link and the page links to separate components.

>     * The logo is not a clickable link, it's just a span

I didn't realize that `VBrand` wasn't a link. Fixed by extracting a component.

>     * The focus ring in the mobile menu includes a thin white ring around the pink one

Fixed.

>     * if you expand the page while the menu is open it stays open. not sure if this is worth fixing. Screenshot: https://share.cleanshot.com/NYBDBV

@sarayourfriend has mentioned it before and I fixed it by adding some CSS to hide the modal on `md` and up. Do you still see this problem with the updated code?

zackkrida reviewed (changes requested):
CSS indeed hides the menu on larger breakpoints now, but the frontend body scrolling remains locked and there remains odd whitespace next to the page content:

![CleanShot 2022-09-26 at 14 02 34@2x](https://user-images.githubusercontent.com/6351754/192348274-c0cdfe55-1503-4c17-be67-a50a520e9101.png)

The focus ring is fixed, except on the WordPress icon:

<img width="499" alt="CleanShot 2022-09-26 at 14 03 49@2x" src="https://user-images.githubusercontent.com/6351754/192348379-b7adfe52-94e8-498c-b81d-39a0eb03ed51.png">


This is so close to merge!

dhruvkb reviewed (approved):
Looks good to me! I noticed no problems in testing and the design matches the mockups.

obulat commented on src/composables/use-pages.ts:
```diff
@@ -7,26 +7,32 @@ export default function usePages(isNew = false) {
     ? [
         {
           id: 'about',
+          name: 'navigation.about',
```
There is a `i18n:check` script in package.json that checks for unused and missing i18n keys. Unfortunately, it can't understand dynamic keys, so it reports the key as missing if it's dynamic. This means that we cannot use it in the CI. If we convert all dynamic keys to static keys, we will be able to check the i18n for completeness in the CI, @dhruvkb

dhruvkb replied:
Ok, makes sense.

zackkrida reviewed (approved):
LGTM!
